### ISSUE TITLE
Separate Redis::connect() from host configuration.

### ISSUE THREAD

**Issue by galves** [phpredis/phpredis#934]
Please consider reworking the Redis::connect API to better support Dependency Injection.  Currently connection configuration must be passed when connect() is called, which results in poor encapsulation.

What I would suggest doing is separating the connection from the configuration so the connection can be established lazily.

Example:

``` php

class RedisCache implements CacheInterface {

    private $redis;

    public function __construct(Redis $client) {
       $this->redis = $client;
    }

   private function getConnection()
   {
       // why should RedisCache care about the server host, it just needs a redis connection?
       $client->isConnected() || $client->connect();
       return $this->redis;
   }

    /** implement CacheInterface *//
}

$redis = new Redis();
$redis->setHost('somehost');
$redis->setPort(6379);

$cache = new RedisCache($redis);

// reuse connection
$logger = new RedisLogger($redis); 

// if logger and cache are never used, no connection will be made.

```

Thanks.

**Comment by bryanlatten** [comment]
I would like to +1, as I convert an old app from memcache --> redis, looking for a replacement for http://php.net/manual/en/memcache.addserver.php, which is described: ```Thus there is no overhead in adding a large number of servers to the pool, even though they might not all be used.``` 

Thanks again for this stable extension, unlike the entire memcache(d) mess!

**Comment by B-Galati** [comment]
Would be really cool to have a lazy connection :+1:

**Comment by B-Galati** [comment]
Still interested for this as it would avoid the usage of dynamic proxy and issues like  snc/SncRedisBundle#559

**Comment by yatsukhnenko** [comment]
We will discuss this issue with @michael-grunder. I have a thoughts to completely remove `connect` method and always make connection lazy

**Comment by michael-grunder** [comment]
I like the idea as well, we just can't do it until a major version bump.

I've created a 6.0 milestone and added this feature to it.  Ideally I would like to do a rather substantial rewrite for PhpRedis 6 with a bunch of big changes, but it will come down to how much time we can reasonably invest in such an endeavor.

**Comment by B-Galati** [comment]
Thanks to both of you :pray:

**Comment by yatsukhnenko** [comment]
@B-Galati @bryanlatten @galves I merged changes to `develop` branch which allows you to set configuration while creating object.

**Comment by bryanlatten** [comment]
@yatsukhnenko are these changes now in mainline?

**Comment by yatsukhnenko** [comment]
The changes are in `develop` branch and will be released with PhpRedis 6.0

**Comment by B-Galati** [comment]
So great! Thanks! 

cc @ostrolucky (https://github.com/phpredis/phpredis/issues/934#issuecomment-599421477)
